### PR TITLE
Enums for NetworkManager and NetworkService  states. JB#61909

### PR DESCRIPTION
**Pull request by pvuorela**
Plus some minor adjustments in separate commits.

### CODE REVIEW

**Comment by pvuorela** [comment]
As commented on the commit message, I'm not entirely happy with the naming, but wanted to keep the current "state" for backward compatibility. Still thinking if there are better options. Maybe NetworkManager::globalState() should be e.g. managerState()?

File: libconnman-qt/networkmanager.cpp [hunk 1]
```diff
@@ -584,7 +576,7 @@ class InterfaceProxy: public QDBusAbstractInterface
 
 public:
     InterfaceProxy(NetworkManager *parent) :
-        QDBusAbstractInterface(CONNMAN_SERVICE, "/", "net.connman.Manager",
+        QDBusAbstractInterface("net.connman", "/", "net.connman.Manager",
```
Comment (LaakkonenJussi) [inline libconnman-qt/networkmanager.cpp]: I know the name for ConnMan D-Bus does not change but do we need to do this change here to drop the one common place to define the name? It seems unnecessary for me to replace the `CONNMAN_SERVICE` in all of these. Why not move that from the deleted to some other header, just for simplicity?

File: libconnman-qt/networkmanager.cpp [hunk 2]
```diff
@@ -1103,12 +1096,13 @@ bool NetworkManager::isAvailable() const
 
 QString NetworkManager::state() const
 {
-    return m_priv->m_propertiesCache[State].toString();
+    // deprecated
```
Comment (LaakkonenJussi) [inline libconnman-qt/networkmanager.cpp]: Shouldn't we add more than a comment for a deprecated feature here?

File: libconnman-qt/networkmanager.h [hunk 3]
```diff
@@ -32,7 +32,9 @@ class NetworkManagerFactory : public QObject
 class NetworkManager : public QObject
 {
     Q_OBJECT
+    Q_ENUMS(State)
     Q_PROPERTY(bool available READ isAvailable NOTIFY availabilityChanged)
+    // deprecated
```
Comment (LaakkonenJussi) [inline libconnman-qt/networkmanager.h:37]: Is there a way to indicate deperacted use for `Q_PROPERTY`?

File: libconnman-qt/networkservice.cpp [hunk 4]
```diff
@@ -724,6 +719,7 @@ inline QString NetworkService::Private::stringValue(const QString &key)
     return m_propertiesCache.value(key).toString();
 }
 
+// deprecated
```
Comment (LaakkonenJussi) [inline libconnman-qt/networkservice.cpp]: Also here, maybe deprecate in away it is visible in some way?

File: libconnman-qt/networkservice.cpp [hunk 5]
```diff
@@ -792,42 +788,45 @@ void NetworkService::Private::setLastConnectError(const QString &error)
...
-        m_state = currentState;
-        queueSignal(SignalStateChanged);
-        updateConnecting(currentState);
-        updateConnected(currentState);
+    m_serviceState = stateStringToEnum(currentState);
+
+    queueSignal(SignalStateChanged);
+    queueSignal(SignalServiceStateChanged);
+
+    if (service()->connecting() != wasConnecting) {
```
Comment (LaakkonenJussi) [inline libconnman-qt/networkservice.cpp:831]: I'm not following now, does the `connecting` and `connected` get changed in between this and the earlier assignment? The signals are queued and this is relying and checking if the signal has been even been processed out of the queue AND something changed on this service due to that? Just wondering about the async nature here in Qt that can it even happen and does the change be done then too late that this state is not updated - or will it be? Again this is a big change...

File: libconnman-qt/networkservice.cpp [hunk 6]
```diff
@@ -969,15 +979,21 @@ void NetworkService::Private::onConnectFinished(QDBusPendingCallWatcher *call)
 
         // InProgress means that somebody has already asked this service
         // to get connected. That's fine, we will keep watching the status.
-        setLastConnectError((errorName == ConnmanError::InProgress) ? QString() : errorName);
+        setLastConnectError(errorName == QStringLiteral("net.connman.Error.InProgress") ? QString() : errorName);
```
Comment (LaakkonenJussi) [inline libconnman-qt/networkservice.cpp]: Maybe we should not hardcode these `"net.connman.Error"`.

File: libconnman-qt/networktechnology.cpp [hunk 7]
```diff
@@ -64,7 +65,7 @@ QSharedPointer<TechnologyTracker> TechnologyTracker::instance()
 
 TechnologyTracker::TechnologyTracker()
     : QObject()
-    , m_dbusWatcher(new QDBusServiceWatcher(CONNMAN_SERVICE, QDBusConnection::systemBus(),
+    , m_dbusWatcher(new QDBusServiceWatcher(QLatin1String("net.connman"), QDBusConnection::systemBus(),
```
Comment (LaakkonenJussi) [inline libconnman-qt/networktechnology.cpp]: Seems like an unnecessary change.

File: libconnman-qt/networktechnology.cpp [hunk 8]
```diff
@@ -81,15 +82,15 @@ TechnologyTracker::TechnologyTracker()
 
     // Monitor TechnologyAdded and TechnologyRemoved.
     QDBusConnection::systemBus().connect(
-                CONNMAN_SERVICE,
+                "net.connman",
```
Comment (LaakkonenJussi) [inline libconnman-qt/networktechnology.cpp]: Seems like an unnecessary change.

File: libconnman-qt/networktechnology.cpp [hunk 9]
```diff
@@ -81,15 +82,15 @@ TechnologyTracker::TechnologyTracker()
...
                 "/",
                 "net.connman.Manager",
                 "TechnologyAdded",
                 this,
                 SLOT(onTechnologyAdded(QDBusObjectPath,QVariantMap)));
 
     QDBusConnection::systemBus().connect(
-                CONNMAN_SERVICE,
+                "net.connman",
```
Comment (LaakkonenJussi) [inline libconnman-qt/networktechnology.cpp]: Seems like an unnecessary change.

File: libconnman-qt/networktechnology.cpp [hunk 10]
```diff
@@ -106,7 +107,7 @@ QSet<QString> TechnologyTracker::technologies()
 
 void TechnologyTracker::getTechnologies()
 {
-    QDBusInterface managerInterface(CONNMAN_SERVICE, "/", "net.connman.Manager", QDBusConnection::systemBus());
+    QDBusInterface managerInterface("net.connman", "/", "net.connman.Manager", QDBusConnection::systemBus());
```
Comment (LaakkonenJussi) [inline libconnman-qt/networktechnology.cpp]: Seems like an unnecessary change.

File: libconnman-qt/networktechnology.cpp [hunk 11]
```diff
@@ -286,7 +287,7 @@ void NetworkTechnology::pendingSetProperty(const QString &key, const QVariant &v
 void NetworkTechnology::createInterface()
 {
     if (!d_ptr->m_path.isEmpty() && d_ptr->m_technologyTracker->technologies().contains(d_ptr->m_path)) {
-        d_ptr->m_technology = new NetConnmanTechnologyInterface(CONNMAN_SERVICE, d_ptr->m_path,
+        d_ptr->m_technology = new NetConnmanTechnologyInterface(QLatin1String("net.connman"), d_ptr->m_path,
```
Comment (LaakkonenJussi) [inline libconnman-qt/networktechnology.cpp]: Seems like an unnecessary change.

File: libconnman-qt/commondbustypes.cpp [hunk 12]
```diff
@@ -8,38 +8,7 @@
  * http://www.apache.org/licenses/LICENSE-2.0
  */
 
-#include "libconnman_p.h"
-
-#define CONNMAN_ERROR "net.connman.Error"
-
-const QString ConnmanError::Failed(CONNMAN_ERROR ".Failed");
```
Comment (LaakkonenJussi) [inline libconnman-qt/commondbustypes.cpp:15]: I'm not sure if it is a good idea to remove these now. I just thought that these could be used more in passing information upwards about what was wrong with the services, and to react to these better. In many ways the error reporting on WLAN and VPN is quite low on detail and user has to guess what was wrong.

**Review by LaakkonenJussi — COMMENTED** [review]
This moves into good direction but unsure if all removals are necessary as of now.

File: libconnman-qt/networkmanager.cpp (hunk 1 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.cpp]: To me it felt a bit pointless having the service name that way while all the other details of the API are hard-coded like the path and the interface. And then one file just for this name would feel like overkill.

Though one middle-ground could be defining it per file.

File: libconnman-qt/networkmanager.h (hunk 3 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.h:37]: Mostly just getter complaining at runtime.

File: libconnman-qt/networkmanager.cpp (hunk 2 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.cpp]: Sure, could add a note there where to go.

File: libconnman-qt/commondbustypes.cpp (hunk 12 above)
Comment (pvuorela) [reply libconnman-qt/commondbustypes.cpp:15]: Can be relatively easily added back if there's something done to this direction. Now it's just dead code.

File: libconnman-qt/networkservice.cpp (hunk 5 above)
Comment (pvuorela) [reply libconnman-qt/networkservice.cpp:831]: It's not really that big of a change. What this is about is that both connecting and connected depend on the state, and thus if the state is changed we need to check if that affected the results of these.

File: libconnman-qt/networkmanager.cpp (hunk 1 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.cpp]: Moved the old definition to commondbustypes.h. Should be close enough to the context.

File: libconnman-qt/networktechnology.cpp (hunk 8 above)
Comment (pvuorela) [reply libconnman-qt/networktechnology.cpp]: Using now the common again on this and later.

File: libconnman-qt/networkservice.cpp (hunk 5 above)
Comment (pvuorela) [reply libconnman-qt/networkservice.cpp:831]: Also to note that the old code did the same checking, it was just having separate methods that were called from this one. What I did there is getting rid of generally redundant member variables for these states.

File: libconnman-qt/networkmanager.cpp (hunk 2 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.cpp]: Added a one-time runtime warning on this and the networkservice case.

File: libconnman-qt/networkservice.cpp (hunk 4 above)
Comment (pvuorela) [reply libconnman-qt/networkservice.cpp]: done.

File: libconnman-qt/networkservice.cpp [hunk 13]
```diff
@@ -792,42 +787,45 @@ void NetworkService::Private::setLastConnectError(const QString &error)
...
+    } else if (state == QStringLiteral("association")) {
+        return NetworkService::AssociationState;
+    } else if (state == QStringLiteral("configuration")) {
+        return NetworkService::ConfigurationState;
+    } else if (state == QStringLiteral("ready")) {
+        return NetworkService::ReadyState;
+    } else if (state == QStringLiteral("disconnect")) {
+        return NetworkService::DisconnectState;
+    } else if (state == QStringLiteral("online")) {
+        return NetworkService::OnlineState;
```
Comment (spiiroin) [inline libconnman-qt/networkservice.cpp:808]: style nit: bunch of else-after-jump things in all new code

File: libconnman-qt/networkservice.cpp [hunk 14]
```diff
@@ -792,42 +787,45 @@ void NetworkService::Private::setLastConnectError(const QString &error)
...
-    QString currentState(state());
-    if (m_state != currentState) {
-        m_state = currentState;
-        queueSignal(SignalStateChanged);
-        updateConnecting(currentState);
-        updateConnected(currentState);
+    m_serviceState = stateStringToEnum(currentState);
+
+    queueSignal(SignalStateChanged);
+    queueSignal(SignalServiceStateChanged);
```
Comment (spiiroin) [inline libconnman-qt/networkservice.cpp:829]: I see no reason why every call to updateState() would mean state change signal is needed. Could be conditional like connected/connecting signals below?

File: libconnman-qt/networkmanager.cpp [hunk 15]
```diff
@@ -14,16 +14,16 @@
 #include <QRegularExpression>
 #include <QWeakPointer>
 
-static const QString InputRequestTimeout("InputRequestTimeout");
+static const QString InputRequestTimeoutProperty("InputRequestTimeout");
```
Comment (spiiroin) [inline libconnman-qt/networkmanager.cpp]: Just wondering if it were better or worse if all property names were grouped together...

File: libconnman-qt/networkmanager.cpp [hunk 16]
```diff
@@ -1515,9 +1512,25 @@ bool NetworkManager::isValid() const
...
+NetworkManager::State NetworkManager::globalState() const
+{
+    QString state = m_priv->m_propertiesCache.value(StateProperty).toString();
+    if (state == QStringLiteral("offline")) {
+        return OfflineState;
+    } else if (state == QStringLiteral("idle")) {
+        return IdleState;
+    } else if (state == QStringLiteral("ready")) {
+        return ReadyState;
+    } else if (state == QStringLiteral("online")) {
```
Comment (spiiroin) [inline libconnman-qt/networkmanager.cpp:1525]: style nit: else-after-jump chain in new code

File: plugin/declarativenetworkmanager.cpp [hunk 17]
```diff
@@ -90,6 +92,11 @@ QString DeclarativeNetworkManager::state() const
     return m_sharedInstance->state();
 }
 
+DeclarativeNetworkManager::State DeclarativeNetworkManager::globalState() const
+{
+    return static_cast<DeclarativeNetworkManager::State>(m_sharedInstance->globalState());
```
Comment (spiiroin) [inline plugin/declarativenetworkmanager.cpp:97]: Random observation: expectation that DeclarativeNetworkManager::State enum matches NetworkManager::State - need for this is documented only at DeclarativeNetworkManager side...

**Review by spiiroin — APPROVED** [review]
Commented about some cosmetic / eye of the beholder things, but did not spot any real issues.

File: libconnman-qt/networkmanager.cpp (hunk 15 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.cpp]: I'll move this a bit.

File: libconnman-qt/networkmanager.cpp (hunk 16 above)
Comment (pvuorela) [reply libconnman-qt/networkmanager.cpp:1525]: As discussed in offline, there are pros and cons, extra lines vs no braces vs this.

File: plugin/declarativenetworkmanager.cpp (hunk 17 above)
Comment (pvuorela) [reply plugin/declarativenetworkmanager.cpp:97]: Was thinking all the places calling this would have state changed but no harm in explicitly guarding that something changed.

File: libconnman-qt/networkservice.cpp (hunk 14 above)
Comment (pvuorela) [reply libconnman-qt/networkservice.cpp:829]: Was thinking all the places calling this would have state changed but no harm in explicitly guarding that something changed.

File: plugin/declarativenetworkmanager.cpp (hunk 17 above)
Comment (pvuorela) [reply plugin/declarativenetworkmanager.cpp:97]: Made the enum values copying NetworkManager side so changes only there have no chance of breaking anything.

File: libconnman-qt/networkservice.cpp (hunk 5 above)
Comment (LaakkonenJussi) [reply libconnman-qt/networkservice.cpp:831]: > It's not really that big of a change. 

I meant this PR as total. But yeah, should be ok.

**Review by LaakkonenJussi — APPROVED** [review]
LGTM, tested also on a device with the other changes, seems to work.

A request: could we keep tagging the commits, also with bug refs, it is much easier to see to which bug the change belongs.
